lib/ledger-node-state-machine.js

```diff
@@ -5,35 +5,79 @@
  */
 module.exports = class LedgerNodeStateMachine {
-  constructor(ledgerNode, storage, consensus) {
+  constructor({ledgerNode}) {
     this.ledgerNode = ledgerNode;
-    this.storage = storage;
-    this.consensus = consensus;
+    this.storage = ledgerNode.storage;
   }
 
   /**
    * Gets an object from the ledger given a objectID and a set of options.
    *
-   * objectId - the URI of the object to fetch.
-   * options - a set of options used when retrieving the object.
-   * callback(err, object) - the callback to call when finished.
+   * recordId - the URI of the object to fetch.
+   * callback(err, record) - the callback to call when finished.
    *   err - An Error if an error occurred, null otherwise.
    */
-  get(objectId, options, callback) {
-    if(typeof options === 'function') {
-      callback = options;
-      options = {};
-    }
+  get({maxBlockHeight, recordId}, callback) {
+    // the history returned from storage will be in the proper order, however
+    // that does not mean all the operations are valid
+    this.storage.operations.getRecordHistory(
+      {maxBlockHeight, recordId}, (err, recordHistory) => {
+        if(err) {
+          return callback(err);
+        }
+        callback(null, _replayHistory({recordHistory}));
+      });
+
+    function _replayHistory({recordHistory}) {
+      // find the first CreateWebLedgerRecord
+      const createIndex = _.findIndex(
+        recordHistory, ['operation.type', 'CreateWebLedgerRecord']);
+      let currentRecord = recordHistory[createIndex].operation.record;
+      const meta = {sequence: 0};
+      if(recordHistory.length === 1) {
+        return {meta, record: currentRecord};
+      }
+      // look for updates after the create operation
+      const updates = [];
+      for(let i = createIndex + 1; i < recordHistory.length; ++i) {
+        if(recordHistory[i].operation.type === 'UpdateWebLedgerRecord') {
+          updates.push(recordHistory[i]);
+        }
+      }
+      for(let i = 0; i < updates.length; ++i) {
+        const {patch, sequence} = updates[i].operation.recordPatch;
+        if(sequence !== meta.sequence) {
+          // skip the update if the sequence is not proper
+          continue;
+        }
+        const errors = jsonpatch.validate(patch, currentRecord);
+        if(errors) {
+          // skip the update if the patch cannot be properly applied
+          continue;
+        }
+        currentRecord = jsonpatch.applyPatch(currentRecord, patch).newDocument;
+        meta.sequence++;
+      }
 
-    // TODO: need to check to see if stateMachine is up-to-date
-    // TODO: if not, need to run StateMutators based on config -- or this
-    //   all needs to happen in a ledger node state machine worker that is
-    //   scheduled similarly to a consensus worker
+      // FIXME: is there a possiblity that the patch added a property that is
+      // not in the context or did the validator eliminate that possiblity?
```

Comment:
The validator handles that. Custom validators could allow properties that aren't in the context but they'll become fully expanded (full URLs).